br/pkg/lightning/backend/external/writer.go

```diff
@@ -252,43 +251,48 @@ type Writer struct {
 	totalSize uint64
 }
 
-// AppendRows appends rows to the external storage.
-// Note that this method is NOT thread-safe.
-func (w *Writer) AppendRows(ctx context.Context, _ []string, rows encode.Rows) error {
-	kvs := kv.Rows2KvPairs(rows)
+// WriteRow implements ingest.Writer.
+func (w *Writer) WriteRow(ctx context.Context, idxKey, idxVal []byte, handle tidbkv.Handle) error {
 	keyAdapter := w.keyAdapter
-	for _, pair := range kvs {
-		w.batchSize += uint64(len(pair.Key) + len(pair.Val))
-
-		buf := w.kvBuffer.AllocBytes(keyAdapter.EncodedLen(pair.Key, pair.RowID))
-		key := keyAdapter.Encode(buf[:0], pair.Key, pair.RowID)
-		val := w.kvBuffer.AddBytes(pair.Val)
-
-		w.writeBatch = append(w.writeBatch, common.KvPair{Key: key, Val: val})
-		if w.batchSize >= w.memSizeLimit {
-			if err := w.flushKVs(ctx, false); err != nil {
-				return err
-			}
+	w.batchSize += uint64(len(idxKey) + len(idxVal))
+
+	var rowID []byte
+	if handle != nil {
+		rowID = handle.Encoded()
+	}
+	buf := w.kvBuffer.AllocBytes(keyAdapter.EncodedLen(idxKey, rowID))
+	key := keyAdapter.Encode(buf[:0], idxKey, rowID)
+	val := w.kvBuffer.AddBytes(idxVal)
+
+	w.writeBatch = append(w.writeBatch, common.KvPair{Key: key, Val: val})
+	if w.batchSize >= w.memSizeLimit {
+		if err := w.flushKVs(ctx, false); err != nil {
+			return err
 		}
 	}
 	return nil
 }
 
-// IsSynced implements the backend.EngineWriter interface.
-func (w *Writer) IsSynced() bool {
-	return false
+// LockForWrite implements ingest.Writer.
+// Since flushKVs is thread-safe in external storage writer,
+// this is implemented as noop.
+func (w *Writer) LockForWrite() func() {
```

Comment:
why does caller need to explictly lock it, rather than `WriteRow` internaly lock the writer?